File: CuponRequest/src/main/java/com/example/CuponRequest/controller/ReqController.java
package com.example.CuponRequest.controller;

import com.example.CuponRequest.devtools.math.MathOperation;
import com.example.CuponRequest.models.Coupon;
import com.example.CuponRequest.req.Req;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.JsonSerializable;
import com.fasterxml.jackson.databind.ObjectMapper;
import okhttp3.Response;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.client.RestTemplate;

import java.io.IOException;
import java.util.Iterator;

@RestController
public class ReqController {
    private final RestTemplate restTemplate;

    public ReqController(RestTemplate restTemplate) {
        this.restTemplate = restTemplate;
    }

    @GetMapping("/test")
    public Coupon Test() throws IOException {
        Response promoResponse = Req.promo_req("flamp");
        Response listResponse = Req.list_req(promoResponse.header("Set-Cookie"));

        assert listResponse.body() != null;
        String jsonListString = listResponse.body().string();
        assert promoResponse.body() != null;
        String jsonPromoString = promoResponse.body().string();

        ObjectMapper objectMapper = new ObjectMapper();
        JsonNode jsonListNode = objectMapper.readTree(jsonListString);
        JsonNode jsonPromoNode = objectMapper.readTree(jsonPromoString);

        JsonNode promoName = jsonListNode.get("data").get("PROMOCODE");
        JsonNode promoStatus = jsonPromoNode.get("data").get("status");
        JsonNode promoMessage = jsonPromoNode.get("data").get("message");
        JsonNode price = jsonListNode.get("data").get("LIST").get(0).get("BASE_PRICE");
        JsonNode priceAfterUserPromo = jsonListNode.get("data").get("LIST").get(0).get("PRICE");


        Coupon promoInfo = new Coupon();
        promoInfo.setPromoName(promoName.toString());
        promoInfo.setStatus(promoStatus.asBoolean());
        promoInfo.setPromoMassage(promoMessage.toString());
        promoInfo.setPromoValue(MathOperation.calculatePercentageChange(price.asDouble(),priceAfterUserPromo.asDouble()));
        return promoInfo;
    }

}

File: CuponRequest/src/main/java/com/example/CuponRequest/controller/ReqController.kt
package com.example.CuponRequest.controller

import com.example.CuponRequest.devtools.math.MathOperation
import com.example.CuponRequest.models.Coupon
import com.example.CuponRequest.req.Req
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.client.RestTemplate
import java.io.IOException

@RestController
class ReqController(private val restTemplate: RestTemplate) {
    @GetMapping("/test")
    @kotlin.Throws(IOException::class)
    fun Test(): Coupon {
        val promoResponse = Req.promo_req("flamp")
        val listResponse = Req.list_req(promoResponse.header("Set-Cookie"))
        val jsonListString = listResponse.body!!.string()
        val jsonPromoString = promoResponse.body!!.string()
        val objectMapper = ObjectMapper()
        val jsonListNode = objectMapper.readTree(jsonListString)
        val jsonPromoNode = objectMapper.readTree(jsonPromoString)
        val promoName = jsonListNode["data"]["PROMOCODE"]
        val promoStatus = jsonPromoNode["data"]["status"]
        val promoMessage = jsonPromoNode["data"]["message"]
        val price = jsonListNode["data"]["LIST"][0]["BASE_PRICE"]
        val priceAfterUserPromo = jsonListNode["data"]["LIST"][0]["PRICE"]
        val promoInfo = Coupon()
        promoInfo.setPromoName(promoName.toString())
        promoInfo.setStatus(promoStatus.asBoolean())
        promoInfo.setPromoMassage(promoMessage.toString())
        promoInfo.setPromoValue(
            MathOperation.calculatePercentageChange(
                price.asDouble(),
                priceAfterUserPromo.asDouble()
            )
        )
        return promoInfo
    }
}